### ISSUE TITLE
Can't convert a function to a trait using an implicit conversion and fully structural types

### ISSUE THREAD

**Issue by hunterpayne** [scala/bug#12609]
## Reproduction steps

Scala version: 2.13.8

```scala
class RS[I]() {

  def map[O](value: ScopedValue[I, O]): RS[O] = ???
}

trait ScopedValue[I, O] { }

object Test {

  implicit def value[I, O](func: (I) => O): ScopedValue[I, O] =
    new ScopedValue[I, O] { }
  
  val exec = new RS[Long]()
  // doesn't compile but should
  val set = exec.map[Long] { l => l }
  // compiles
  val set2 = exec.map[Long] { (l: Long) => l }
}

```

## Problem

It seems that the function needs to be fully specified to make the implicit conversion happen.  Otherwise you get an 'missing parameter type' error which seems wrong.  If the spec says that this is the correct behavior please let me know and I will close this.

**Comment by Jasper-M** [comment]
The tag says "fixed in Scala 3" but it's not, as far as I can see. Also, I'm no type inference expert, but it seems unlikely that this will ever work.

**Comment by joroKr21** [comment]
Right - we have to know what is the type of the thing we convert before we convert it

**Comment by joroKr21** [comment]
Similar issue with extension methods: [lampepfl/dotty/issues/15427](https://github.com/lampepfl/dotty/issues/15427)

**Comment by hunterpayne** [comment]
But the method is requesting a ScopedValue[I, O] which can be constructed using the implicit conversion of a (I) => O to a ScopedValue[I, O] (hence the structural type).  So because the I comes from exec and is specified (it is a Long), the compiler has all the information needed to determine that l => l is l: Long => l.

BTW, is this really fixed in Dotty/Scala 3?  It would be a lot of work to upgrade to Scala 3 but if that is the fix then I suppose it is worth it.

**Comment by SethTisue** [comment]
> is this really fixed in Dotty/Scala 3? 

sorry about that, not sure what I was thinking, I tried it again just now and Scala 3.1.3-RC4 does give:

```
15 |  val set = exec.map[Long] { l => l }
   |                             ^
   |                          Missing parameter type
```

**Comment by SethTisue** [comment]
The spec has relatively little to say about type inference.

But I suspect it's as simple @joroKr21 said:

> we have to know what is the type of the thing we convert before we convert it

**Comment by som-snytt** [comment]
I sympathize with the OP.

The handling of under-specified function literals for overloading (https://dotty.epfl.ch/docs/reference/changed-features/overload-resolution.html) suggests (to a certain naive belief) that the same should be possible when conversions are available.

Take my literal as `? => ?`, the expected type is `ScopedValue[Long, Long]`, then the expected type for the conversion is `Long => Long`.

**Comment by hunterpayne** [comment]
If I change (I) => O to something that isn't a function, this code compiles.  So I feel strongly that it should work.  It is only the raising of the passed type to a function that makes this not work.

**Comment by hunterpayne** [comment]
```
class RS[I]() {

  def map[O](value: ScopedValue[I, O]): RS[O] = ???
}

trait ScopedValue[I, O] { }

object Test {

  implicit def value[I, O](func: (I) => O): ScopedValue[I, O] =
    new ScopedValue[I, O] { }
  implicit def value2[I, O](func: I): ScopedValue[I, O] =
    new ScopedValue[I, O] { }
  
  val exec = new RS[Long]()
  // doesn't compile but should
  val set = exec.map[Long] { l => l }
  // compiles
  val set2 = exec.map[Long] { l: Long => l }
  // compiles
  val set3 = exec.map[Long] { 7L }
}
```

**Comment by Jasper-M** [comment]
> ```scala
> val set3 = exec.map[Long] { 7L }
> ```

But here you have a value of type `Long` while in the other you have a value of type `Function1[?, ?]` where there are still unresolved type variables. 

I think a more similar example would be
```scala
class Foo[A, B]
implicit def value2[I, O](foo: Foo[I, O]): ScopedValue[I, O] =
    new ScopedValue[I, O] { }

//doesn't compile
exec.map[Long] { new Foo }
```

**Comment by hunterpayne** [comment]
In that case B is never specified.  Let me try again

```

class RS[I]() {

  def map[O](value: ScopedValue[I, O]): RS[O] = ???
}

class Foo[I, O]() {
}

trait ScopedValue[I, O] { }

object Test {

  implicit def value[I, O](func: (I) => O): ScopedValue[I, O] =
    new ScopedValue[I, O] { }
  implicit def value2[I](func: Seq[I]): ScopedValue[I, Long] =
    new ScopedValue[I, Long] { }
  implicit def value3[I](func: Foo[I, Long]): ScopedValue[I, Long] =
    new ScopedValue[I, Long] { }

  val exec = new RS[Long]()
  // doesn't compile but should
  // val set = exec.map[Long] { l => l }
  // compiles
  val set2 = exec.map[Long] { l: Long => l }
  // compiles
  val set3 = exec.map[Long] { Seq() }
  // doesn't compile
  val set4 = exec.map[Long] { new Foo }
}
```

So it appears that if 2 type parameters are present it doesn't work but if only 1 is present, it does?

**Comment by Jasper-M** [comment]
Something else is at play here. In `exec.map[Long] { Seq() }` the compiler still can't infer the type parameter to `Seq.apply` so he defaults to `Nothing`, just like with `Foo`. But because `Seq` is covariant in its type parameter, `Seq[Nothing]` is a subtype of `Seq[Long]` and `value2` becomes an eligible implicit conversion.

**Comment by hunterpayne** [comment]
So you are right that this compiles:
```
class RS[I]() {

  def map[O](value: ScopedValue[I, O]): RS[O] = ???
}

class Foo[+I, +O]() {
}

trait ScopedValue[I, O] { }

object Test {

  implicit def value[I, O](func: (I) => O): ScopedValue[I, O] =
    new ScopedValue[I, O] { }
  implicit def value2[I](func: Seq[I]): ScopedValue[I, Long] =
    new ScopedValue[I, Long] { }
  implicit def value3[I](func: Foo[I, Long]): ScopedValue[I, Long] =
    new ScopedValue[I, Long] { }

  val exec = new RS[Long]()
  // doesn't compile but should
  // val set = exec.map[Long] { l => l }
  // compiles
  val set2 = exec.map[Long] { l: Long => l }
  // compiles
  val set3 = exec.map[Long] { Seq() }
  // doesn't compile
  val set4 = exec.map[Long] { new Foo }
}
```

However, Function1 is contravariant in the first type parameter.  So it still doesn't compile for the case I really care about.  It is also weird that it requires covariance as it is increasingly difficult to use covariance and contravariance in Scala.  You can't even have an implicit TypeTag of a covariant or contravariant type which makes using them very difficult.  In my use case, it is a complete non-starter as I need the type tag to do what we require.  So back to square 1.

**Comment by hunterpayne** [comment]
> Right - we have to know what is the type of the thing we convert before we convert it

I do want to address this before I give up hope here.  The implicit conversion in question can convert any type of Function1, it has no restrictions.  So no, you are wrong and you do not have to know the type of the thing first, you only have to know that it is a Function1 which it clearly is.  And half of the required type is specified even before you do the implicit conversion as the argument is known to be a ScopedValue[Long, _] beforehand because exec is of type RS[Long].  If you do the implicit conversions, the compiler would find that the rest of the type is indeed specified by the return value of the function.  BTW, this compiler error happens even when the return value of the function doesn't depend on the argument type (e.g. j => 7L).  I used this specific example because perhaps l => l is a good corner case to cover too.

This seems like an issue that has been reported several times by others so I guess you will never fix this issue.  Or maybe you just can't.

**Comment by som-snytt** [comment]
Probably no one read my previous comment, but I suggested "doable".

What you would sacrifice is some implicit scope, which would normally include the type args of the `Function[A, R]`.

Intuitively, this is user-defined SAM conversion, where the target type is not a SAM type, but the context knows how to wire the function to the type.

My comment about overloading was intended to say that overloading used to break inference in a similar way. Both have function literals in need of an expected type.

I don't know if it would be too much complexity for whatever the use case is.

**Comment by hunterpayne** [comment]
So the use case would be something like, "using an under-specified implicit conversion to infer Function type parameters".  And the repo case is the original one for this bug.  The type inference in this case stops when it finds an unspecified function when what it should do is pull in I from the function call and O from the return type of the Function1[I, O].  Is this correct?

**Comment by hunterpayne** [comment]
I created a unit test to verify this issue.  It simply tries to compile the repo case above.  And I added this line to the type inference code:
```      Console.println(s"typed ${tree} in mode ${mode} pt ${pt}")```
on line 6011 of Typers.scala.

Here is the output I get from the case that works (where the type of the anonymous function's argument is specified):

```
typed exec.map[Long](((l: Long) => l)) in mode EXPRmode pt ?
typed exec.map[Long] in mode APPSELmode-BYVALmode-EXPRmode-FUNmode-POLYmode pt 
typed exec.map in mode APPSELmode-BYVALmode-EXPRmode-FUNmode-POLYmode-TAPPmode pt ?
typed exec in mode APPSELmode-EXPRmode-POLYmode-QUALmode pt ?
typed Long in mode TYPEmode pt ?
typed ((l: Long) => l) in mode BYVALmode-EXPRmode pt ScopedValue[Long,Long]
```

And here is the output when I try the repo case (where the type of the anonymous function isn't specified):

```
typed exec.map[Long](((l) => l)) in mode EXPRmode pt ?
typed exec.map[Long] in mode APPSELmode-BYVALmode-EXPRmode-FUNmode-POLYmode pt 
typed exec.map in mode APPSELmode-BYVALmode-EXPRmode-FUNmode-POLYmode-TAPPmode pt ?
typed exec in mode APPSELmode-EXPRmode-POLYmode-QUALmode pt ?
typed Long in mode TYPEmode pt ?
typed ((l) => l) in mode BYVALmode-EXPRmode pt ScopedValue[Long,Long]
```

So the type inference engine is able to find the correct type with no changes (nice).  Somehow the compiler error is thrown incorrectly further along in the process.  So it seems to me that the fix is to somehow add a check to the code that throws the compiler error to avoid this case.  However, I don't want to start commenting out lines that throw the compiler error farther down the line and I don't know the correct checks to add.  Some help determining how to construct the proper check would be appreciated.

**Comment by hunterpayne** [comment]
Can someone trigger the travis jobs for this PR.  Been waiting for almost a week to run the CI jobs.

**Comment by som-snytt** [comment]
@hunterpayne I'd suggest squashing and `push -f` and maybe also add `[ci: last-only]` to the PR title (IIRC). That last-only command reduces the chance that a run doesn't sync. Also, there a sync command for comments which Seth has used effectively, or perhaps ineffectively.

**Comment by hunterpayne** [comment]
just made a new PR: https://github.com/scala/scala/pull/10067

**Comment by hunterpayne** [comment]
I don't know if this matter but in TravisCI it says
`
        We are unable to start your build at this time. You exceeded the number of users allowed for your plan. Please review your plan details and follow the steps to resolution.
`

**Comment by SethTisue** [comment]
That's probably coming from your fork? In the main repo, https://github.com/scala/scala/pull/10067 already shows a green Travis-CI run.

**Comment by hunterpayne** [comment]
You are probably right but the validate-main job still hasn't run.

**Comment by hunterpayne** [comment]
The builds are all complete and successful.  Thank you for your help.

**Comment by SethTisue** [comment]
(Jenkins had a transient disk-space issue; Lukas fixed it.)

**Comment by SethTisue** [comment]
as per discussion on https://github.com/scala/scala/pull/10067, we don't have consensus that there is a bug here. hence, closing
